### PR TITLE
[TD] added TechDrawTools Commands

### PR DESCRIPTION
**Pull request by edi271**
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [ ]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [ ]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [ ]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the [FreeCAD bug tracker issue number](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) in case a particular commit solves or is related to an existing issue on the tracker. Ex: `Draft: fix typos - fixes #0004805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

### CODE REVIEW

File: src/Mod/TechDraw/Gui/CommandCreateDims.cpp [hunk 1]
```diff
@@ -602,6 +602,7 @@ CmdTechDrawHorizontalDimension::CmdTechDrawHorizontalDimension()
 void CmdTechDrawHorizontalDimension::activated(int iMsg)
 {
     Q_UNUSED(iMsg);
+    QMessageBox::warning(Gui::getMainWindow(),QObject::tr("Gestartet"),QObject::tr("Horizontal dimension"));
```
Comment (luzpaz) [inline src/Mod/TechDraw/Gui/CommandCreateDims.cpp:605]: What is 'Gestartet' ?

**Comment by luzpaz** [comment]
'Command tools' seems very generic to me. That description could mean anything in my mind. Maybe we should name it something a little more indicative of what the tools do?

File: src/Mod/TechDraw/Gui/CommandCreateDims.cpp (hunk 1 above)
Comment (edi271) [reply src/Mod/TechDraw/Gui/CommandCreateDims.cpp:605]: Sorry, this statement should be commented. It's just needed during development. 'Gestartet' means 'started'

**Comment by yorikvanhavre** [comment]
This looks good ideed, but I agree with @luzpaz , maybe a better name than "Tools"?  Also, these should be documented. I reckon there is no forum thread related to these changes? Would you create the appropriate doc pages on the wiki?

**Comment by edi271** [comment]
> This looks good ideed, but I agree with @luzpaz , maybe a better name than "Tools"? Also, these should be documented. I reckon there is no forum thread related to these changes? Would you create the appropriate doc pages on the wiki?

Please give me a proposal how to name the command file. I am open for any name. The fundament of the suggested changes in TechDraw is the macro TechDrawTools.FCMacro https://github.com/FreeCAD/FreeCAD-macros/blob/master/TechDraw/TechDrawTools.FCMacro It contains 38 tools divided into three groups, doing very different things. My aim is to create all tools in C++. I am a beginner and therefore try to realize the project in small portions. In this PR you find the first one.

There is a forum thread https://forum.freecadweb.org/viewtopic.php?f=17&t=60488 related to the suggested changes in TechDraw. Also there is a second thread related to the macro which is the pattern https://forum.freecadweb.org/viewtopic.php?f=35&t=59334 In this second thread you will find a documentation created by @domad, using animated gifs.

I would like to create appropriate doc pages on the wiki, and I hope @domad will help me. I will contact him. But till now I have no licence to create wiki pages. Must the wiki pages been created before the changes can be merged ?

**Comment by luzpaz** [comment]
TechDraw Extension Pack?  
@edi271 contact me on https://gitter.im/luzpaz with a valid email and I can set you up with wiki access :wink:

**Comment by luzpaz** [comment]
@pavltom care to weigh in ?

**Comment by yorikvanhavre** [comment]
Screw Mark Tools? Mark Tools? Marking Tools? Centermark Tools? Extra Markings? I would go with something generic as it would be more flexible and allows to extend late on...

**Comment by aapo-aapo** [comment]
> Screw Mark Tools? Mark Tools? Marking Tools? Centermark Tools? Extra Markings? I would go with something generic as it would be more flexible and allows to extend late on...

Could it perhaps be "Cosmetic Tools", as all of these tools add so-called Cosmetic Arcs, Cosmetic Circles, Cosmetic Lines, and Cosmetic Points.  That way it would be possible to add other tools later on under the same umbrella.  I think it was originally WandererFan that named this stuff "Cosmetic".

**Comment by yorikvanhavre** [comment]
Cosmetic tools seems very good to me!

**Comment by aapo-aapo** [comment]
> Cosmetic tools seems very good to me!

Yes, the good thing would certainly be that it'd be in line with WandererFan's original creation of "Cosmetic Vertices" etc.  On the other hand, "Cosmetic" is a bit of a jargon term, and gives a bit of a beauty salon feeling (which might not be bad, after all!).  I'd still really like those to be called "Auxiliary Lines" and this new feature to be "Auxiliary Marking Tools" or something similar, but it might be too late to change "Cosmetic" to "Auxiliary" TechDraw-wise.  And I'm sure WandererFan had his reasons.  I also like your suggestion "Marking Tools", maybe that would be easiest to understand for users, even though it wouldn't directly signify that Cosmetic Objects will be added by pressing this?

**Comment by MarcelP102** [comment]
perhaps it's better to call it "Annotation tools"?

**Comment by aapo-aapo** [comment]
> perhaps it's better to call it "Annotation tools"?

I believe the term "Annotation" in TechDraw is currently used for text annotations (The icon with the "A" letter), so users would probably think that it'd be related to editing text input on the drawings.

**Comment by edi271** [comment]
> > perhaps it's better to call it "Annotation tools"?
> 
> I believe the term "Annotation" in TechDraw is currently used for text annotations (The icon with the "A" letter), so users would probably think that it'd be related to editing text input on the drawings.

Hello to all:

This PR is just the start of a bigger project. My aim is to implement all of the more then 30 tools of TechDrawTools.FCMacro. Don't know how much success I will have.

Meanwhile I cleared the faults reported in https://forum.freecadweb.org/viewtopic.php?f=17&t=60488

I propose to call it CommandExtensionPack.cpp. If you give me an OK to this name I will update the PR. Or give me another name.

**Comment by yorikvanhavre** [comment]
OK for me :) You could then call the toolbar "Extension Pack¨... Seems appropriate

**Comment by edi271** [comment]
Hello to all:

This PR is just the start of a bigger project. My aim is to implement all of the more then 30 tools of TechDrawTools.FCMacro. Don't know how much success I will have.

Meanwhile I cleared the faults reported in https://forum.freecadweb.org/viewtopic.php?f=17&t=60488

I propose to call it CommandExtensionPack.cpp. If you give me an OK to this name I will update the PR. Or give me another name.

**Comment by aapo-aapo** [comment]
> This PR is just the start of a bigger project. My aim is to implement all of the more then 30 tools of TechDrawTools.FCMacro. Don't know how much success I will have.

Hmm, there are going to be quite many new tools, then.  Nice!  I noticed that you have divided them in your Macro into 3 categories on different tabs: "Attributes and Modifications", "Centerlines and Threading", and "Dimensions".  Would it be too much work to keep the grouping of the tools similarly separate on the c++ side, too?

So, your current PR appears to add tools from "Centerlines and Threading" ("Extra Geometry"?) , and you could perhaps later add "Dimension Helpers" and "Attributes and Modifications" as separate toolsets with new PR:s.  This way you could have, e.g. "ExtraGeometry.cpp", "DimensionHelpers.cpp", and "AttributeModifications.cpp"; of which you'd now add just the first file.

I'd think that at least keeping the UI commands handling Dimensions and Cosmetic Geometry separate would make the code in TechDraw cleaner and easier to understand.

**Comment by balrobs** [comment]
Just compiled the latest code and redid my testings. I now can now confirm that from a user perspective everything works fine!

**Comment by yorikvanhavre** [comment]
Ok, thanks for having adapted the code, @edi271 and thanks everybody else for testing! This now looks perfect to me too, so let's merge!

**Pull request merged**